File: app/src/main/java/com/numero/sojodia/model/TimeTableRow.java
package com.numero.sojodia.model;

import android.annotation.SuppressLint;

@SuppressLint("DefaultLocale")
public class TimeTableRow {

    private String hourText;

    private StringBuilder timeOnWeekdayStringBuilder = new StringBuilder();
    private StringBuilder timeOnSaturdayStringBuilder = new StringBuilder();
    private StringBuilder timeOnSundayStringBuilder = new StringBuilder();

    public TimeTableRow() {
    }

    public void setHour(int hour) {
        this.hourText = String.format("%02d", hour);
    }

    public String getHourText() {
        return hourText;
    }

    public void addTimeOnWeekday(int time, boolean isNonStop) {
        if (isNonStop) {
            timeOnWeekdayStringBuilder.append("★");
        }
        timeOnWeekdayStringBuilder.append(String.format("%02d ", time));
    }

    public String getTimeOnWeekdayText() {
        return timeOnWeekdayStringBuilder.toString();
    }

    public void addTimeOnSaturday(int time, boolean isNonStop) {
        if (isNonStop) {
            timeOnSaturdayStringBuilder.append("★");
        }
        timeOnSaturdayStringBuilder.append(String.format("%02d ", time));
    }

    public String getTimeOnSaturdayText() {
        return timeOnSaturdayStringBuilder.toString();
    }

    public void addTimeOnSunday(int time, boolean isNonStop) {
        if (isNonStop) {
            timeOnSundayStringBuilder.append("★");
        }
        timeOnSundayStringBuilder.append(String.format("%02d ", time));
    }

    public String getTimeOnSundayText() {
        return timeOnSundayStringBuilder.toString();
    }
}

File: app/src/main/kotlin/com/numero/sojodia/model/TimeTableRow.kt
package com.numero.sojodia.model

import java.util.*

data class TimeTableRowList(
        val value: List<TimeTableRow>
) {

    companion object {
        fun mapToTimeTableRowList(busTimeList: List<BusTime>): TimeTableRowList {
            val firstHour = busTimeList.first().time.hour
            val lastHour = busTimeList.last().time.hour
            val busTimeGroup = busTimeList
                    .asSequence()
                    .groupBy { it.time.hour }
            val list = (firstHour..lastHour).map { hour ->
                TimeTableRow(hour).apply {
                    busTimeGroup[hour]?.forEach { busTime ->
                        when (busTime.week) {
                            Week.WEEKDAY -> addBusTimeOnWeekday(busTime)
                            Week.SATURDAY -> addBusTimeOnSaturday(busTime)
                            Week.SUNDAY -> addBusTimeOnSunday(busTime)
                        }
                    }
                }
            }
            return TimeTableRowList(list)
        }
    }
}

data class TimeTableRow(val hour: Int) {

    private val onWeekdayBusTime: MutableList<BusTime> = mutableListOf()
    private val onSaturdayBusTime: MutableList<BusTime> = mutableListOf()
    private val onSundayBusTime: MutableList<BusTime> = mutableListOf()

    fun createOnWeekdayText(isNotSchoolTerm: Boolean): String {
        return onWeekdayBusTime.createBusMinTimeString(isNotSchoolTerm)
    }

    fun createOnSaturdayText(isNotSchoolTerm: Boolean): String {
        return onSaturdayBusTime.createBusMinTimeString(isNotSchoolTerm)
    }

    fun createOnSundayText(isNotSchoolTerm: Boolean): String {
        return onSundayBusTime.createBusMinTimeString(isNotSchoolTerm)
    }

    fun addBusTimeOnWeekday(busTime: BusTime) {
        onWeekdayBusTime.add(busTime)
    }

    fun addBusTimeOnSaturday(busTime: BusTime) {
        onSaturdayBusTime.add(busTime)
    }

    fun addBusTimeOnSunday(busTime: BusTime) {
        onSundayBusTime.add(busTime)
    }

    private fun List<BusTime>.createBusMinTimeString(isNotSchoolTerm: Boolean): String {
        return buildString {
            this@createBusMinTimeString.asSequence().forEach {
                if (isNotSchoolTerm.not() or it.isOnlyOnSchooldays.not()) {
                    if (it.isNonstop) {
                        append("★")
                    }
                    append(String.format(Locale.ENGLISH, "%02d ", it.time.min))
                }
            }
        }
    }
}
